python/revng/cli/translate/pipelines/isolate-translate.yml

```diff
@@ -5,7 +5,12 @@ Steps:
    Pipes:
      - Name:             LLVMPipe
        UsedContainers: [module]
-       Passes: [detect-abi, isolate, invoke-isolated-functions]
+       Passes: [collect-functions-from-callees,
+                detect-abi,
+                collect-functions-from-unused-addresses,
+                detect-abi,
```

Comment:
I noticed that abi detection pass is called twice. Is that an intentional decision? Why can’t we just detect all the functions before running it only once? Does the second level of detection depends on abi or something?

cc: @aleclearmind

Reply:
`detect-abi` goals is to obtain a prototype for all the function in the model that have an `Invalid` function type.
The idea is that you:
1. do a first pass (`collect-functions-from-callees`) collecting function looking at `call` instructions (which is a very reliable way to identify functions) ;
2. use `detect-abi` to discover their prototype (and their CFG)
3. then you do another pass (`collect-functions-from-unused-addresses`) which tries to identify other functions that are not target of direct calls (e.g., addresses we found in global data, which is typical for e.g., virtual tables). The reason for splitting these two collections is that, in this second phase, we ignore all the function entry points the we have previously detect to be part of another function.

Basically the idea is: try to find addresses that we're highly confident represent a function entry point, identify their CFG, and then have a less reliable source of function entry points which we filter in order to exclude entry points that, at the previous iteration, we identified as part of another function (and therefore it's likely to be a false positive).

Makes sense?

Reply:
So what you are saying is that my suspicion about
```
Passes: [collect-functions-from-callees, collect-functions-from-unused-addresses, detect-abi]
```
begin equivalent to
```
Passes: [collect-functions-from-callees, detect-abi, collect-functions-from-unused-addresses, detect-abi]
```
is not wrong - it's just that we want to "mark" them different for now so that at some point we might want to process them differently. Am I understanding it right?

Reply:
If you run `detect-abi` only once at the end, `collect-functions-from-unused-addresses` will not see the CFGs of the functions identified by `collect-functions-from-callees` and will not be able to blacklist any candidate, even if it will later be detect as part of another function.

But now that I think about it, since we only need the CFG, instead of running `detect-abi`, we could run only the part recovering the CFG.

Reply:
Ah, wait, CFG was only recovered as a part of abi-detection? I had no idea :D